### PR TITLE
[Merged by Bors] - refactor(Dual): drop a `DecidableEq` assumption

### PR DESCRIPTION
**Pull request by urkud**
Split `DualBases.eval` into `eval_same` and `eval_of_ne` to avoid a `DecidableEq` assumption.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/Dual.lean [hunk 1]
```diff
@@ -682,10 +682,10 @@ elab "use_finite_instance" : tactic => evalUseFiniteInstance
 /-- `e` and `ε` have characteristic properties of a basis and its dual -/
 -- @[nolint has_nonempty_instance] Porting note (#5171): removed
 structure Module.DualBases (e : ι → M) (ε : ι → Dual R M) : Prop where
-  eval : ∀ i j : ι, ε i (e j) = if i = j then 1 else 0
+  eval_same : ∀ i, ε i (e i) = 1
+  eval_of_ne : Pairwise fun i j ↦ ε i (e j) = 0
```
Comment (urkud) [inline Mathlib/LinearAlgebra/Dual.lean:686]: Another option is to use `eval : ∀ [DecidableEq ι], ∀ i j ...`

File: Mathlib/LinearAlgebra/Dual.lean [hunk 2]
```diff
@@ -755,17 +755,14 @@ def basis : Basis ι R M :=
...
   ext i
   rw [Basis.apply_eq_iff]
   ext j
-  rw [h.basis_repr_apply, coeffs_apply, h.eval, Finsupp.single_apply]
-  convert if_congr (eq_comm (a := j) (b := i)) rfl rfl
+  rcases eq_or_ne i j with rfl | hne
+  · simp [h.eval_same]
+  · simp [hne, h.eval_of_ne hne.symm]
 
 -- `convert` to get rid of a `DecidableEq` mismatch
```
Comment (Command-Master) [inline Mathlib/LinearAlgebra/Dual.lean:767]: Is this comment still relevant?

Comment (grunweg) [reply Mathlib/LinearAlgebra/Dual.lean:767]: I don't see any `convert` anywhere close to this - can you remove this comment, please?
```suggestion

**Review by grunweg — COMMENTED** [review]
Thanks! Let's get another pair of eyes on this.
maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**